### PR TITLE
Remove global const  in init in cholmod

### PR DESCRIPTION
**Pull request by musm**
cc @yuyichao @vtjnash

### CODE REVIEW

**Review by yuyichao — COMMENTED** [review]
You should probably make `commonStruct` an actual const (resize/fill on init) and others const `Ref`s to avoid performance regression.

File: base/sparse/cholmod.jl [hunk 1]
```diff
@@ -89,10 +88,9 @@ function __init__()
                 of CHOLMOD, or download the generic binaries
                 from www.julialang.org, which ship with the correct
                 versions of all dependencies.
-            """)
```
Comment (fredrikekre) [inline base/sparse/cholmod.jl]: Why delete the empty lines and add these spaces? Not super important, but I think the previous behavior is nicer.

**Review by yuyichao — APPROVED** [review]
The change itself LGTM. (I did not check if all users are corrected).

File: base/sparse/cholmod.jl [hunk 2]
```diff
@@ -125,27 +133,28 @@ function __init__()
...
 
         ### Initiate CHOLMOD
         ### The common struct. Controls the type of factorization and keeps pointers
         ### to temporary memory.
-        global const commonStruct = fill(0xff, common_size)
+        resize!(commonStruct, common_size)
+        fill!(commonStruct, 0xff)
 
-        global const common_supernodal =
+        common_supernodal[] =
```
Comment (yuyichao) [inline base/sparse/cholmod.jl]: You can probably remove the `convert`.

Comment (musm) [reply base/sparse/cholmod.jl]: Yeah. 
Thank you for the review and suggested improvements!

File: base/sparse/cholmod.jl [hunk 3]
```diff
@@ -1330,7 +1338,7 @@ function cholfact!(F::Factor{Tv}, A::Sparse{Tv}; shift::Real=0.0) where Tv
     cm = common()
```
Comment (musm) [inline base/sparse/cholmod.jl]: Is it really necessary to have a function `common()` to get the global variable  `commonStruct`?

Comment (andreasnoack) [reply base/sparse/cholmod.jl]: No, I don't think so. It used to take an argument but that was broken and got removed so I think we can just delete the function.

Comment (musm) [reply base/sparse/cholmod.jl]: done

**Comment by musm** [comment]
This should be g2g now. perhaps run benchmarks?

**Comment by musm** [comment]
https://github.com/JuliaLang/julia/blob/master/test/codegen.jl#L51 
https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.18278/job/tbwdkqhcbqtngayu#L6034
codegen test failure on x86_64 for reference (which does not seem to be at all related to the changes here)

**Comment by fredrikekre** [comment]
Looks like #23365

File: base/sparse/cholmod.jl (hunk 3 above)
Comment (musm) [reply base/sparse/cholmod.jl]: @andreasnoack would you be opposed to renaming `commonStruct` to `common_struct` ? The cammelcase doesn't seem very commonly used in julia code.

**Comment by musm** [comment]
tests all pass sans an unrelated error on appveyor. Should we run nano soldier here just to double check?

**Comment by andreasnoack** [comment]
@nanosoldier `runbenchmarks("linalg", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23449#issuecomment-325159944) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3969d6fca3b100bed48c43101bf132f883435190/7457905_vs_ab0ac5e/report.md). cc @ararslan

**Comment by musm** [comment]
G2G?

**Comment by musm** [comment]
@fredrikekre why? It wasn't four space aligned .....

**Comment by fredrikekre** [comment]
The string was 4 space indented, you removed one space making the string 5 spaces indented so I added it back :)

**Comment by musm** [comment]
Please see line 100 and 116, it's now misaligned with those

**Comment by fredrikekre** [comment]
CI passed on the previous commit so should be ok to squash merge, and remove the ci skip in the process.

**Comment by fredrikekre** [comment]
That's not the point, the point is to have the indentation correct in the string, see https://docs.julialang.org/en/stable/manual/strings/#Triple-Quoted-String-Literals-1

**Comment by ViralBShah** [comment]
I think we should not have indentation discussions. Or the author of any PR should be free to ignore anyone who points it out (unless they like the suggestion).

**Pull request merged**